components/playlist/renderer/playlist_render_frame_observer.cc

```diff
@@ -60,4 +65,69 @@ void PlaylistRenderFrameObserver::HideMediaSourceAPI() {
       blink::WebString::FromUTF16(kScriptToHideMediaSourceAPI)));
 }
 
+void PlaylistRenderFrameObserver::InstallMediaDetector() {
+  DVLOG(2) << __FUNCTION__;
+
+  static const char16_t kScriptToDetectVideoAndAudio[] =
+      uR"-(
+    (function() {
+      // Firstly, we try to get find all <video> or <audio> tags periodically,
+      // for a a while from the start up. If we find them, then we attach 
+      // MutationObservers to them to detect source URL.
+      // After a given amount of time, we do this in requestIdleCallback().
+      // Note that there's a global object named |pl_worker|. This worker is
+      // created and bound by PlaylistJSHandler.
+
+      let mutationSources = new Set();
+      const findNewMediaAndObserveMutation = () => {
+          return document.querySelectorAll('video, audio').forEach((mediaNode) => {
+              if (mutationSources.has(mediaNode)) return
+
+              mutationSources.add(mediaNode)
+              pl_worker.onMediaUpdated(mediaNode.src)
+
+              new MutationObserver(mutations => {
+                mutations.forEach(mutation => { pl_worker.onMediaUpdated(mutation.target.src); })
```

Comment:
If we're not sharing the observer between multiple elements we know exactly what changed here so we could (optionally) do this instead:

```suggestion
                pl_worker.onMediaUpdated(mediaNode.src);
```